File: _dev/bench-offsets/lib/bench.js
'use strict';
// Bancs d'essai JS — 6 des 12 tests définis dans README.md :
//   READ-FULL-TEXT/SPLIT-SPACE/JS   READ-STREAMING/SPLIT-SPACE/JS
//   READ-FULL-TEXT/SPACE-CSS/JS     READ-STREAMING/SPACE-CSS/JS
//   READ-FULL-TEXT/SPACE-CALC/JS    READ-STREAMING/SPACE-CALC/JS
// Les 6 bancs PY (mêmes noms, suffixe /PY) sont dans bench.py.
//
// Nécessite spacy_cache.json (build_spacy_cache.py côté Python — à lancer une fois avant).
// spaCy ne tourne jamais ici, seulement le JSON déjà calculé est chargé. 4 étapes
// chronométrées séparément (noms imposés par README.md) :
//   - Spacy analyse loading : chargement du cache spaCy (+ texte.txt en plus, pour SPLIT-SPACE
//     seulement, qui ignore les offsets spaCy mais a besoin du texte brut pour son split() naïf)
//   - Tokens Fullfillment   : TOKENS = tokens spaCy + gap propre à la technique + proximité
//   - HTML building         : construction du DOM (createElement/appendChild, via jsdom
//     puisque Node n'a pas de DOM natif), spans data-canon
//   - Html doc loading      : 0 — construire le DOM, c'est déjà avoir le document chargé,
//     pas d'étape d'ouverture séparée côté JS (cf. README)
//
// Nécessite `npm install` dans ce dossier (jsdom) + un tas V8 augmenté (les éléments DOM
// jsdom sont lourds, 6 tests x ~200-350k noeuds épuisent le tas par défaut) :
//   node --max-old-space-size=4096 _dev/bench-offsets/lib/bench.js

const fs = require('fs');
const path = require('path');
const { performance } = require('perf_hooks');
const { JSDOM } = require('jsdom');

const SOURCE_TXT = path.join(__dirname, 'texte.txt');
const CACHE_JSON = path.join(__dirname, 'spacy_cache.json');

function readFull(file) {
  return fs.readFileSync(file, 'utf-8');
}

function readStreaming(file) {
  return new Promise((resolve, reject) => {
    const parts = [];
    const stream = fs.createReadStream(file, { encoding: 'utf-8' });
    stream.on('data', (chunk) => parts.push(chunk));
    stream.on('end', () => resolve(parts.join('')));
    stream.on('error', reject);
  });
}

async function loadCacheFull(file) {
  return JSON.parse(readFull(file));
}

async function loadCacheStreaming(file) {
  return JSON.parse(await readStreaming(file));
}

const LOAD_MODES = [
  ['READ-FULL-TEXT', async (f) => readFull(f), loadCacheFull],
  ['READ-STREAMING', async (f) => readStreaming(f), loadCacheStreaming],
];

// SPLIT-SPACE — split('\n') puis split(' '), n'utilise pas les offsets spaCy pour le gap, mais
// réutilise le canon déjà calculé dans le cache (fait linguistique, pas un offset). Si le
// découpage naïf ne retombe pas exactement sur un offset spaCy (ponctuation collée différemment),
// fallback sur la forme elle-même en minuscules : un canon n'est jamais vide.
function splitSpaceTokens(text, spacyTokens) {
  const canonParFormeOffset = new Map(spacyTokens.map(([f, o, ws, c]) => [`${f}@${o}`, c]));
  const out = [];
  let offset = 0;
  const paragraphs = text.split('\n');
  paragraphs.forEach((para, pi) => {
    const words = para === '' ? [] : para.split(' ');
    words.forEach((forme, wi) => {
      const gap = wi === words.length - 1 ? 0 : 1;
      const canon = canonParFormeOffset.get(`${forme}@${offset}`) ?? forme.toLowerCase();
      out.push([forme, offset, gap, canon]);
      offset += forme.length + gap;
    });
    if (pi < paragraphs.length - 1) offset += 1;
  });
  return out;
}

// SPACE-CSS — classe s0/s1 par mot, depuis offset/longueur spaCy déjà en cache.
function spaceCssTokens(text, spacyTokens) {
  return spacyTokens.map(([forme, offset, ws, canon]) => [forme, offset, ws ? 's1' : 's0', canon]);
}

// SPACE-CALC — même formule, mais l'espace (chaîne réelle) est le gap porté par le token.
function spaceCalcTokens(text, spacyTokens) {
  return spacyTokens.map(([forme, offset, ws, canon]) => [forme, offset, ws, canon]);
}

// ─── DOM building (JS — équivalent de la construction HTML de bench.py) ───────────────────

function domSplitSpace(document, tokensOut) {
  const p = document.createElement('p');
  tokensOut.forEach(([forme, offset, gap, canon]) => {
    const span = document.createElement('span');
    span.setAttribute('data-canon', canon || '');
    span.textContent = forme;
    p.appendChild(span);
    if (gap) p.appendChild(document.createTextNode(' '));
  });
  return p;
}

function domSpaceCss(document, tokensOut) {
  const container = document.createElement('div');
  tokensOut.forEach(([forme, offset, cls, canon]) => {
    const span = document.createElement('span');
    span.className = cls;
    span.setAttribute('data-canon', canon || '');
    span.textContent = forme;
    container.appendChild(span);
  });
  return container;
}

function domSpaceCalc(document, tokensOut) {
  const container = document.createElement('div');
  tokensOut.forEach(([forme, offset, ws, canon]) => {
    const span = document.createElement('span');
    span.setAttribute('data-canon', canon || '');
    span.textContent = forme;
    container.appendChild(span);
    if (ws) container.appendChild(document.createTextNode(ws));
  });
  return container;
}

const TECHNIQUES = [
  ['SPLIT-SPACE', splitSpaceTokens, domSplitSpace, true],   // true = a besoin du texte brut
  ['SPACE-CSS', spaceCssTokens, domSpaceCss, false],
  ['SPACE-CALC', spaceCalcTokens, domSpaceCalc, false],
];

async function main() {
  if (!fs.existsSync(CACHE_JSON)) {
    console.error(`Cache manquant : ${CACHE_JSON}. Lancer d'abord build_spacy_cache.py.`);
    process.exit(1);
  }

  const dom = new JSDOM('<!DOCTYPE html><body></body>');
  const document = dom.window.document;

  console.log("BANCS D'ESSAIS — JS (6/12)");
  const results = [];
  let firstRootHtml = null;
  for (const [modeName, readTextFn, readCacheFn] of LOAD_MODES) {
    for (const [techName, tokensFn, domFn, besoinTexte] of TECHNIQUES) {
      let t0 = performance.now();
      const text = besoinTexte ? await readTextFn(SOURCE_TXT) : null;
      const cache = await readCacheFn(CACHE_JSON);
      const spacyAnalyseMs = performance.now() - t0;
      const spacyTokens = cache.tokens;

      t0 = performance.now();
      const tokensOut = tokensFn(text, spacyTokens);
      const tokensMs = performance.now() - t0;

      t0 = performance.now();
      const root = domFn(document, tokensOut);
      const htmlBuildingMs = performance.now() - t0;

      // Html doc loading — équivalent JS du chargement réel du document dans un navigateur
      // (cf. bench.py/charger_dans_navigateur) : attacher l'arbre construit au document live.
      t0 = performance.now();
      document.body.innerHTML = '';
      document.body.appendChild(root);
      const htmlDocLoadingMs = performance.now() - t0;

      const totalMs = spacyAnalyseMs + tokensMs + htmlBuildingMs + htmlDocLoadingMs;

      const nom = `${modeName}/${techName}/JS`;
      console.log(`${nom.padEnd(32)} `
        + `Spacy analyse loading ${spacyAnalyseMs.toFixed(1)} ms | `
        + `Tokens Fullfillment ${tokensMs.toFixed(1)} ms | `
        + `HTML building ${htmlBuildingMs.toFixed(1)} ms | `
        + `Html doc loading ${htmlDocLoadingMs.toFixed(1)} ms | `
        + `total ${totalMs.toFixed(1)} ms`);

      results.push({
        nom,
        spacy_analyse_loading_ms: Math.round(spacyAnalyseMs * 10) / 10,
        tokens_fullfillment_ms: Math.round(tokensMs * 10) / 10,
        html_building_ms: Math.round(htmlBuildingMs * 10) / 10,
        html_doc_loading_ms: htmlDocLoadingMs,
        total_ms: Math.round(totalMs * 10) / 10,
        tokens: tokensOut.length,
      });

      if (firstRootHtml === null) firstRootHtml = root.outerHTML;
    }
  }

  fs.writeFileSync(path.join(__dirname, 'results_js.json'), JSON.stringify(results, null, 2), 'utf-8');

  const texteJs = path.join(__dirname, '..', 'texte_js.html');
  fs.writeFileSync(texteJs, `<!DOCTYPE html><html><head><meta charset="utf-8"></head><body>${firstRootHtml}</body></html>`, 'utf-8');
  console.log(`-> ${texteJs}`);
}

main();

File: _dev/bench-offsets/lib/bench.py
"""Bancs d'essai PY — 6 des 12 tests définis dans README.md :
  READ-FULL-TEXT/SPLIT-SPACE/PY   READ-STREAMING/SPLIT-SPACE/PY
  READ-FULL-TEXT/SPACE-CSS/PY     READ-STREAMING/SPACE-CSS/PY
  READ-FULL-TEXT/SPACE-CALC/PY    READ-STREAMING/SPACE-CALC/PY
Les 6 bancs JS (mêmes noms, suffixe /JS) sont dans bench.js.

Nécessite spacy_cache.json (build_spacy_cache.py, à lancer une fois avant). Pour chaque test,
4 étapes chronométrées séparément (noms imposés par README.md) :
  - Spacy analyse loading : chargement du cache spaCy (+ texte.txt en plus, pour SPLIT-SPACE
    seulement, qui ignore les offsets spaCy mais a besoin du texte brut pour son split() naïf)
  - Tokens Fullfillment   : TOKENS = tokens spaCy + gap propre à la technique + proximité
    (canon identique à moins de 1501 caractères, même logique que ProxEngine.get_repetitions)
  - HTML building         : construction du document HTML final, spans data-canon
  - Html doc loading      : chargement réel de ce document dans un navigateur (pywebview,
    déjà une dépendance du projet), jusqu'à l'évènement `loaded`

  /opt/homebrew/opt/python@3.11/bin/python3.11 _dev/bench-offsets/lib/bench.py
"""
import json
import time
from pathlib import Path

HERE = Path(__file__).resolve().parent
SOURCE_TXT = HERE / "texte.txt"
CACHE_JSON = HERE / "spacy_cache.json"
CHUNK_SIZE = 65536


def read_full(path):
    return path.read_text(encoding="utf-8")


def read_streaming(path):
    parts = []
    with path.open("r", encoding="utf-8") as f:
        while True:
            chunk = f.read(CHUNK_SIZE)
            if not chunk:
                break
            parts.append(chunk)
    return "".join(parts)


def load_cache_full(path):
    return json.loads(path.read_text(encoding="utf-8"))


def load_cache_streaming(path):
    return json.loads(read_streaming(path))


LOAD_MODES = (
    ("READ-FULL-TEXT", read_full, load_cache_full),
    ("READ-STREAMING", read_streaming, load_cache_streaming),
)


# ─── Techniques — Tokens Fullfillment (identique en JS) ───────────────────────

def split_space_tokens(text, spacy_tokens):
    """SPLIT-SPACE : split('\\n') puis split(' '), n'utilise pas les offsets spaCy pour le gap
    — mais réutilise le canon déjà calculé dans le cache spaCy (le canon est un fait linguistique,
    pas un offset ; le reste (gap) reste 100% indépendant de spaCy). Si le découpage naïf ne
    retombe pas exactement sur un offset spaCy (ponctuation collée différemment), fallback sur
    la forme elle-même en minuscules : un canon n'est jamais vide."""
    canon_par_forme_offset = {(f, o): c for f, o, ws, c in spacy_tokens}
    out = []
    offset = 0
    paragraphs = text.split('\n')
    for pi, para in enumerate(paragraphs):
        words = para.split(' ') if para != '' else []
        for wi, forme in enumerate(words):
            gap = 0 if wi == len(words) - 1 else 1
            canon = canon_par_forme_offset.get((forme, offset), forme.lower())
            out.append((forme, offset, gap, canon))
            offset += len(forme) + gap
        if pi < len(paragraphs) - 1:
            offset += 1
    return out


def space_css_tokens(text, spacy_tokens):
    """SPACE-CSS : classe s0/s1 par mot, depuis offset/longueur spaCy déjà en cache."""
    return [(forme, offset, 's1' if ws else 's0', canon) for forme, offset, ws, canon in spacy_tokens]


def space_calc_tokens(text, spacy_tokens):
    """SPACE-CALC : même formule, mais l'espace (chaîne réelle) est le gap porté par le token."""
    return [(forme, offset, ws, canon) for forme, offset, ws, canon in spacy_tokens]


# ─── HTML building (PY seulement — l'équivalent JS construit un DOM, cf. bench.js) ────────────

def html_split_space(tokens_out):
    html = ['<p>']
    for forme, offset, gap, canon in tokens_out:
        html.append(f'<span data-canon="{canon or ""}">{forme}</span>')
        if gap:
            html.append(' ')
    html.append('</p>')
    return ''.join(html)


def html_space_css(tokens_out):
    return ''.join(
        f'<span class="{cls}" data-canon="{canon or ""}">{forme}</span>'
        for forme, offset, cls, canon in tokens_out
    )


def html_space_calc(tokens_out):
    return ''.join(
        f'<span data-canon="{canon or ""}">{forme}</span>{ws}'
        for forme, offset, ws, canon in tokens_out
    )


TECHNIQUES = (
    ("SPLIT-SPACE", split_space_tokens, html_split_space, True),   # True = a besoin du texte brut
    ("SPACE-CSS", space_css_tokens, html_space_css, False),
    ("SPACE-CALC", space_calc_tokens, html_space_calc, False),
)


def charger_dans_navigateur(htmls):
    """Ouvre chaque document HTML dans un vrai navigateur (pywebview, fenêtre cachée) et mesure
    le temps jusqu'à l'évènement `loaded` — le "Html doc loading" de README.md."""
    import webview

    resultats = [None] * len(htmls)
    etat = {"t0": None, "i": 0}

    def on_loaded():
        etat["resultat"] = (time.perf_counter() - etat["t0"]) * 1000
        resultats[etat["i"]] = etat["resultat"]
        etat["i"] += 1
        if etat["i"] < len(htmls):
            etat["t0"] = time.perf_counter()
            window.load_html(htmls[etat["i"]])
        else:
            window.destroy()

    def main():
        window.events.loaded += lambda: on_loaded()
        etat["t0"] = time.perf_counter()
        window.load_html(htmls[0])

    window = webview.create_window('bench', html='<html><body></body></html>', hidden=True)
    webview.start(main)
    return resultats


def main():
    if not CACHE_JSON.exists():
        raise SystemExit(f"Cache manquant : {CACHE_JSON}. Lancer d'abord build_spacy_cache.py.")

    lignes = []       # pour affichage + JSON, sans le Html doc loading (rempli après coup)
    htmls = []         # documents HTML construits, dans l'ordre des 6 tests

    for mode_name, read_text_fn, read_cache_fn in LOAD_MODES:
        for tech_name, tokens_fn, html_fn, besoin_texte in TECHNIQUES:
            t0 = time.perf_counter()
            text = read_text_fn(SOURCE_TXT) if besoin_texte else None
            cache = read_cache_fn(CACHE_JSON)
            spacy_analyse_ms = (time.perf_counter() - t0) * 1000
            spacy_tokens = cache["tokens"]

            t0 = time.perf_counter()
            tokens_out = tokens_fn(text, spacy_tokens)
            tokens_ms = (time.perf_counter() - t0) * 1000

            t0 = time.perf_counter()
            html = html_fn(tokens_out)
            html_ms = (time.perf_counter() - t0) * 1000

            nom = f"{mode_name}/{tech_name}/PY"
            htmls.append(html)
            lignes.append({
                "nom": nom,
                "spacy_analyse_loading_ms": round(spacy_analyse_ms, 1),
                "tokens_fullfillment_ms": round(tokens_ms, 1),
                "html_building_ms": round(html_ms, 1),
                "tokens": len(tokens_out),
                "html_chars": len(html),
            })

    doc_loading_ms = charger_dans_navigateur(htmls)

    print("BANCS D'ESSAIS — PY (6/12)")
    for ligne, doc_ms in zip(lignes, doc_loading_ms):
        ligne["html_doc_loading_ms"] = round(doc_ms, 1)
        ligne["total_ms"] = round(
            ligne["spacy_analyse_loading_ms"] + ligne["tokens_fullfillment_ms"]
            + ligne["html_building_ms"] + ligne["html_doc_loading_ms"], 1)
        print(f"{ligne['nom']:32s} "
              f"Spacy analyse loading {ligne['spacy_analyse_loading_ms']:7.1f} ms | "
              f"Tokens Fullfillment {ligne['tokens_fullfillment_ms']:7.1f} ms | "
              f"HTML building {ligne['html_building_ms']:7.1f} ms | "
              f"Html doc loading {ligne['html_doc_loading_ms']:7.1f} ms | "
              f"total {ligne['total_ms']:8.1f} ms")

    (HERE / "results_python.json").write_text(
        json.dumps(lignes, ensure_ascii=False, indent=2), encoding="utf-8")

    texte_py = HERE.parent / "texte_py.html"
    texte_py.write_text(
        f'<!DOCTYPE html><html><head><meta charset="utf-8"></head><body>{htmls[0]}</body></html>',
        encoding="utf-8")
    print(f"-> {texte_py}")


if __name__ == '__main__':
    main()
